python/ray/dag/dag_node.py

```diff
@@ -335,6 +354,33 @@ def __call__(self, node: "DAGNode"):
             )
         )
 
+    def bfs(self, fn: "Callable[[DAGNode], T]"):
+        """Breadth-first search on the DAG."""
+        visited = set()
+        queue = [self]
+        while queue:
+            node = queue.pop(0)
+            if node not in visited:
+                fn(node)
+                visited.add(node)
+                # Add all unseen downstream and upstream nodes to the queue.
+                # This function should be called by the root of the DAG. However,
+                # in some invalid cases, some nodes may not be descendants of the
+                # root. Therefore, we also add upstream nodes to the queue so that
+                # a meaningful error message can be raised when the DAG is compiled.
+                for node in node._downstream_nodes + node._upstream_nodes:
```

Comment:
nit; but list + list op copies the entire list. Can you instead use `from itertools import chain`

Reply:
Updated https://github.com/ray-project/ray/pull/47261/commits/54df435b62b4510d731bd9df4b766f694e54d5ea